includes/class-llms-events.php

```diff
@@ -0,0 +1,411 @@
+<?php
+/**
+ * LifterLMS Event management.
+ *
+ * @package  LifterLMS/Classes
+ *
+ * @since [version]
+ * @version [version]
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * LLMS_Events class.
+ *
+ * @since [version]
+ */
+class LLMS_Events {
+
+	/**
+	 * Singleton instance
+	 *
+	 * @var  null
+	 */
+	protected static $_instance = null;
+
+	/**
+	 * Get Main Singleton Instance.
+	 *
+	 * @since [version]
+	 *
+	 * @return LLMS_Events
+	 */
+	public static function instance() {
+		if ( is_null( self::$_instance ) ) {
+			self::$_instance = new self();
+		}
+		return self::$_instance;
+	}
+
+	/**
+	 * Private Constructor.
+	 *
+	 * @since [version]
+	 *
+	 * @return void
+	 */
+	private function __construct() {
+
+		add_action( 'init', array( $this, 'store_cookie' ) );
+
+	}
+
+	/**
+	 * Retrieves an array of client settings used to initialize the JS Tracking instance on the frontend.
+	 *
+	 * @since [version]
+	 *
+	 * @return array
+	 */
+	public function get_client_settings() {
+
+		$events = ! $this->should_track_client_events() ? array() : array_keys( array_filter( $this->get_registered_events() ) );
+
+		/**
+		 * Filter client-side tracking settings
+		 *
+		 * @since [version]
+		 *
+		 * @param array $settings {
+		 *     Hash of client-side settings.
+		 *
+		 *     @type string $nonce Nonce used to verify client-side events.
+		 *     @type string[] $events Array of events that should be tracked.
+		 * }
+		 */
+		return apply_filters(
+			'llms_events_get_client_settings',
+			array(
+				'nonce'  => wp_create_nonce( 'llms-tracking' ),
+				'events' => $events,
+			)
+		);
+
+	}
+
+	/**
+	 * Retrieve an array of valid events.
+	 *
+	 * @since [version]
+	 *
+	 * @return array Array key is the event name and array value is used to determine if the key is a client-side event.
+	 */
+	public function get_registered_events() {
+
+		$events = array(
+			'account.signon'  => false,
+			'account.signout' => false,
+			'session.start'   => false,
+			'session.end'     => false,
+			'page.load'       => true,
+			'page.exit'       => true,
+			'page.focus'      => true,
+			'page.blur'       => true,
+		);
+
+		/**
+		 * Filter the list of registered events.
+		 *
+		 * Allows 3rd parties to register (or unregister) tracked events.
+		 *
+		 * @since [version]
+		 *
+		 * @param array $events Array of events. Array key is the event name and array value is used to determine if the key is a client-side event.
+		 */
+		return apply_filters( 'llms_get_registered_events', $events );
```

Comment:
this filter hook can be fired many times, what if we remove it and instead:
1) create an `$events` property.
2) we populate it the first time `get_registered_events()` is called (or better when `$this->events` is not set), and there filter the events with a `llms_registered_events` filter hook.
3) any other call to `get_registered_events()` will use the already set `$this->events` without triggering any filter hook.
?

Reply:
Smart, update .